record_archiver/models/record_lifespan.py

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+# © 2015-2016 Yannick Vaucher (Camptocamp SA)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import logging
+
+from datetime import datetime
+from dateutil.relativedelta import relativedelta
+
+from openerp import api, exceptions, fields, models
+from openerp.tools.translate import _
+
+_logger = logging.getLogger(__name__)
+
+
+class RecordLifespan(models.Model):
+    """ Configure records lifespans per model
+
+    After the lifespan is expired (compared to the `write_date` of the
+    records), the records are deactivated.
+    """
+    _name = 'record.lifespan'
+    _order = 'model'
+
+    model_id = fields.Many2one(
+        'ir.model',
+        string='Model',
+        required=True,
+        domain=[('has_an_active_field', '=', True)],
+    )
+    model = fields.Char(
+        related='model_id.model',
+        string='Model Name',
+        store=True,
+    )
+    months = fields.Integer(
+        required=True,
+        help="Number of month after which the records will be set to inactive "
+             "based on their write date"
+    )
+
+    _sql_constraints = [
+        ('months_gt_0', 'check (months > 0)',
+         "Months must be a value greater than 0"),
+    ]
+
+    @api.model
+    def _scheduler_archive_records(self):
+        lifespans = self.search([])
+        _logger.info('Records archiver starts archiving records')
+        for lifespan in lifespans:
+            try:
+                lifespan.archive_records()
+            except exceptions.UserError as e:
+                _logger.error("Archiver error:\n%s", e[1])
+        _logger.info('Rusty Records now rest in peace')
+        return True
+
+    @api.multi
+    def _archive_domain(self, expiration_date):
+        """ Returns the domain used to find the records to archive.
+
+        Can be inherited to change the archived records for a model.
+        """
+        model = self.env[self.model_id.model]
+        domain = [('write_date', '<', expiration_date)]
+        if 'state' in model._columns:
+            domain += [('state', 'in', ('done', 'cancel'))]
```

Comment:
No, the state needs one of the 2 states to work. If one of the two is missing, it doesn't matter. There is no check on the domain compared values.
This is the default behaviour as `'done'` and `'cancel'` are standard states.

If you made a model than doesn't respect this, you can still override `_archive_domain` in a new module to extend this function.